### PR TITLE
feat(composite): support nested update & set in updateOne

### PR DESCRIPTION
**Pull request by Weakky**
## Overview

- Adds support for `set` & `update` on `updateOne` operation.

## Example

```prisma
model Test {
    id Int @id
    to_one ToOne
}

type ToOne {
    field Int
    nested_to_one NestedToOne
}

type NestedToOne {
  field String
}
```

```graphql
mutation {
  updateOneTest(
    where: { id: 1 }      
      data: {
        to_one: { update: { field: 1 }}
  # or: to_one: { update: { field: { set|increment|decrement|...: 1 } } }
  # or: to_one: { update: { nested_to_one: { update: { field: "test" } } } }
  # or: to_one: { update: { nested_to_one: { set: { field: "test" } } } }
  # or: to_one: { update: { nested_to_one: { field: "test" } } }
      }
  ) { id }
}
```

### CODE REVIEW

File: query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    if let WriteExpression::NestedWrite(nested_write) = write_expr {
+        let docs = unfold_nested_write(field, nested_write, &mut vec![field.db_name().to_owned()])
+            .into_iter()
+            // TODO: figure out why we can't flat_map here
+            // TODO: the trait `FromIterator<Vec<mongodb::bson::Document>>` is not implemented for `std::result::Result<Vec<mongodb::bson::Document>, MongoError>`
+            .map(|(write_expr, field, field_name)| render_update_docs(write_expr, field, &field_name))
+            .collect::<crate::Result<Vec<_>>>()?;
+
+        return Ok(docs.into_iter().flatten().collect_vec());
+    };
```
Comment (Weakky) [inline query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs:22]: Do you have any idea what's up here?

File: query-engine/core/src/query_graph_builder/write/write_args_parser.rs [hunk 2]
```diff
@@ -24,49 +26,13 @@ impl WriteArgsParser {
...
-
-                            args.args.insert(sf, expr)
-                        }
-                        _ => unreachable!(),
-                    },
+                    Field::Scalar(sf) if sf.is_list() => {
+                        let expr = parse_scalar_list(v)?;
 
+                        args.args.insert(sf, expr);
+                    }
```
Comment (Weakky) [inline query-engine/core/src/query_graph_builder/write/write_args_parser.rs:33]: I couldn't split this whole bit into a reusable function in this end. The reason is `Field::Relation` which wouldn't return a `WriteExpression`

File: query-engine/core/src/query_graph_builder/write/write_args_parser.rs [hunk 3]
```diff
@@ -111,26 +55,133 @@ impl WriteArgsParser {
...
 fn is_composite_envelope(map: &ParsedInputMap) -> bool {
-    // [Composites] Flavian Todo
-    map.iter()
-        .next()
-        .map(|(key, _)| map.len() == 1 && key == operations::SET)
-        .unwrap_or(false)
+    let envelope_keys = [operations::SET, operations::UPDATE];
+
+    map.iter().any(|(k, _)| envelope_keys.contains(&k.as_str()))
 }
```
Comment (Weakky) [inline query-engine/core/src/query_graph_builder/write/write_args_parser.rs:136]: This is expected to be refactored once we add the flags on envelope input types

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs [hunk 4]
```diff
@@ -203,3 +196,119 @@ fn update_operations_object_type(
...
+
+/// Builds the `update` input object type. Should be used in the envelope type.
+fn composite_update_object_type(ctx: &mut BuilderContext, cf: &CompositeFieldRef) -> InputObjectTypeWeakRef {
+    let name = format!("{}UpdateInput", cf.typ.name);
+
+    let ident = Identifier::new(name, PRISMA_NAMESPACE);
+    return_cached_input!(ctx, &ident);
+
+    let mut input_object = init_input_object_type(ident.clone());
+    input_object.set_min_fields(1);
```
Comment (Weakky) [inline query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:258]: Do we agree on the `set_min_fields(1)` for this object type? This is to prevent:

```graphql
mutation { updateOneX(data: {
  cf_field: { update: {} <-- this empty object }
}) } {
  id
}
```

File: query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs (hunk 1 above)
Comment (Weakky) [reply query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs:22]: Additional note: this whole file is refactored in a follow-up PR.

Comment (dpetrick) [reply query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs:22]: Check out the signature of `flat_map`: In essence, it tells you that instead of returning single elements in `map`, flat map returns an `IntoIterator` that will be consumed to flatten. 

`render_update_docs` returns a `Result` and result types in rust are `IntoIterator`! They return 0 elements if they are `Err`, 1 if they are `OK`, so writing `.flat_map(|(write_expr, field, field_name)| render_update_docs(write_expr, field, &field_name))` is actually fine. What is NOT fine is then trying to collect into `Result<Vec<_>>`. Why? Because the result is already consumed during the flattening because internally, `flat_map` calls `into_iter` and handles the magic for you. Since you don't have `Result` anymore, you _could_ collect like this: ` .collect::<Vec<_>>();`, but note with this approach you're losing all error information as errors are simply zero-length iterators. 

Tl;dr: You can't use `flat_map` here because that would lose error info. You need a two-step approach. You could also rewrite it into a for loop and not iterate twice.

File: query-engine/core/src/query_graph_builder/write/write_args_parser.rs [hunk 5]
```diff
@@ -116,18 +60,126 @@ fn is_composite_envelope(map: &ParsedInputMap) -> bool {
...
         operations::SET => WriteExpression::Value(value.try_into()?),
+        operations::UPDATE => parse_composite_updates(cf, value.try_into()?, path)?,
+        // operations::PUSH => WriteExpression::Add(value.try_into()?),
         _ => unimplemented!(),
     };
 
     Ok(expr)
 }
 
+fn parse_composite_updates(
```
Comment (dpetrick) [inline query-engine/core/src/query_graph_builder/write/write_args_parser.rs:154]: Much better

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/create.rs [hunk 6]
```diff
@@ -126,7 +126,7 @@ impl DataInputFieldMapper for CreateDataInputFieldMapper {
         // If the composite field in _not_ on a model, then it's nested and we're skipping the create envelope for now.
         // (This allows us to simplify the parsing code for now.)
         let mut input_types = if cf.container().as_model().is_some() {
-            vec![shorthand_type.clone(), envelope_type]
+            vec![envelope_type, shorthand_type.clone()]
```
Comment (dpetrick) [inline query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/create.rs:129]: Any particular reason for the ordering?

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs [hunk 7]
```diff
@@ -141,28 +141,21 @@ impl DataInputFieldMapper for UpdateDataInputFieldMapper {
...
-                Arc::downgrade(&input_object)
-            }
-        };
+        if cf.is_list() {
+            input_types.push(InputType::list(shorthand_type));
+        }
 
-        input_field("foo", InputType::object(obj), None)
+        input_field(cf.name.clone(), input_types, None)
+            .nullable_if(!cf.is_required())
```
Comment (dpetrick) [inline query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:157]: I'm not 100% sure right now, but I believe this must also check for `!cf.is_list()`. I'm not sure how lists and field requiredness work together off the top of my head.

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs [hunk 8]
```diff
@@ -203,3 +196,120 @@ fn update_operations_object_type(
 
     Arc::downgrade(&obj)
 }
+
+// Builds a `set` input object type. Should be used for the shorthand syntax and within the update envelope type.
+fn composite_set_update_object_type(ctx: &mut BuilderContext, cf: &CompositeFieldRef) -> InputObjectTypeWeakRef {
```
Comment (dpetrick) [inline query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:201]: I'm pretty sure this is unnecessary. How would this ever differ from the set input that is created for the create operations?

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs (hunk 4 above)
Comment (dpetrick) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:259]: A quick experiment shoes me that we don't do that for other updates as well. Is this important? We could just treat it as noop like relations.

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs [hunk 9]
```diff
@@ -203,3 +196,120 @@ fn update_operations_object_type(
...
+// Builds an `unset` input field. Should only be used in the envelope type.
+fn composite_unset_update_input_field(cf: &CompositeFieldRef) -> Option<InputField> {
+    if cf.is_required() || cf.is_list() {
+        return None;
+    }
+
+    Some(input_field(operations::UNSET, InputType::boolean(), None).optional())
+}
+
+// Builds an `update` input field. Should only be used in the envelope type.
```
Comment (dpetrick) [inline query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:292]:
```suggestion
// Builds a `set` input field. Should only be used in the envelope type.
```

Comment (dpetrick) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:292]: Is this correct?

**Review by dpetrick — APPROVED** [review]

File: query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs (hunk 1 above)
Comment (Weakky) [reply query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs:22]: Oh lord, that wasn't trivial. Super useful explanation, thank you 🙏. I'll refactor that to a for-loop in the follow-up PR if you don't mind 👍

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs (hunk 4 above)
Comment (Weakky) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:259]: Yeah, I noticed it wasn't the case on existing `update`s. I thought this might remove potential ambiguities where users might think `{ a: { update: {} } }` would update `a` to `{}`

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/create.rs (hunk 6 above)
Comment (Weakky) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/create.rs:129]: For the playground to pick up the envelope types instead of the shorthand ones. Makes it a lot easier to navigate the API.

File: query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+use itertools::Itertools;
+use mongodb::bson::{doc, Document};
+use prisma_models::PrismaValue;
+
+pub fn render_update_docs(
+    write_expr: WriteExpression,
+    field: &Field,
+    field_name: &str,
+) -> crate::Result<Vec<Document>> {
+    if let WriteExpression::NestedWrite(nested_write) = write_expr {
```
[the remark this answers is not in the document]
Comment (Weakky) [reply query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs:13]: It is indeed 👍

File: query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+            "$set": { field_name: { "$divide": [dollar_field_name, (field, rhs).into_bson()?] } }
+        },
+        WriteExpression::NestedWrite(_) => unreachable!(),
+        WriteExpression::Field(_) => unimplemented!(),
+    };
+
+    Ok(vec![doc])
+}
+
+fn unfold_nested_write<'a>(
```
[the remark this answers is not in the document]
Comment (Weakky) [reply query-engine/connectors/mongodb-query-connector/src/root_queries/update_utils.rs:97]: You saw the other PR already, which already includes a comment. I'll make sure to better explain the data transformation there.

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs (hunk 9 above)
Comment (Weakky) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:292]: good catch, thanks!

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs (hunk 8 above)
Comment (Weakky) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:201]: s t u p i d 🤦‍♂️

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs (hunk 7 above)
Comment (Weakky) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:157]: Yeah, you're right. I added an `is_optional` helper that checks for `FieldArity::Optional`.

File: query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs (hunk 4 above)
Comment (Weakky) [reply query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs:259]: Decision: keep the `min(1)`.
